syntax = "proto3";

message getRequestClient {
	uint32 key = 1;
	string consistencyLevel = 2;
}

message putRequestClient {
	uint32 key = 1;
	string value = 2;
	string consistencyLevel = 3;
} 

message coordinatorResponseMessage {
	string value = 1;
}

message coordinatorExceptionMessage {
	string errorMessage = 1;
}

message getMessage {
	uint32 key = 1;
}

message putMessage {
	uint32 key = 1;
	string value = 2;
	string timestamp = 3;
}

message returnPutMessage {
	string messageString = 1;
}

message returnGetMessage {
	string timestamp = 1;
	string value = 2;
	bool notFound = 3;  	// Key not found
}

message replicaInitMessage {
	string ipAddress = 1;
	uint32 portNumber = 2;
}

message ReplicaMessage {
	oneof replica_message {
		getRequestClient get_request = 1;
		putRequestClient put_request = 2;
		coordinatorResponseMessage coordinator_response = 3;
		coordinatorExceptionMessage coordinator_exception = 4;
		getMessage get_message = 5;
		putMessage put_message = 6;
		returnPutMessage return_put_message = 7;
		returnGetMessage return_get_message = 8;
		replicaInitMessage replica_init_message = 9;
	}
	string src_replicaID = 10;
	string dst_replicaID = 11;
}
